src/Microsoft.Extensions.Http/Logging/HttpHeadersLogValue.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Net.Http.Headers;
+using System.Text;
+
+namespace Microsoft.Extensions.Http.Logging
+{
+    internal class HttpHeadersLogValue : IReadOnlyList<KeyValuePair<string, object>>
+    {
+        private readonly Kind _kind;
+
+        private string _formatted;
+        private List<KeyValuePair<string, object>> _values;
+
+        public HttpHeadersLogValue(Kind kind, HttpHeaders headers, HttpHeaders contentHeaders)
+        {
+            _kind = kind;
+
+            Headers = headers;
+            ContentHeaders = contentHeaders;
+        }
+
+        public HttpHeaders Headers { get; }
+
+        public HttpHeaders ContentHeaders { get; }
+        
+        private List<KeyValuePair<string, object>> Values
+        {
+            get
+            {
+                if (_values == null)
+                {
+                    var values = new List<KeyValuePair<string, object>>();
```

Comment:
We don't really have a choice but to allocate. The `HttpHeaders` types are pretty opaque.

This only happens at Trace anyway